### DISCUSSION TITLE
How to create a gap between code and table when rendering to docx ?

### DISCUSSION THREAD

**Discussion by Andrzej-Andrzej** [quarto-dev/quarto-cli#5442] [Q&A]
Hi,
How to increase the space between the code and its output (table, graph ) in MS Word rendered document ? With default rendering it produces no space, code and resulting table looks merged together in MS Word. How to increase that distance, meaning no space with that grey code area  and table placed underneath that code exists - how to separate this ?
Additionally I have got resulting  table squeezed in Word and I must manually adjust width of the columns as columns' names are divided between two consecutive lines ?
Plaease help me with this. Thank you.

**Comment by Andrzej-Andrzej** [comment]
Actually to be more precise I want to createa a space between output (of the first line of code in a chunk) and next, new line of code. 
I found something here:
[https://stackoverflow.com/questions/74575072/insert-pagebreak-for-output-after-code-chunk](url)
but that is creating a break after code chunk. So how to adapt this to my needs, please ?

**Comment by mcanouil** [comment]
Could you share a small working (reproducible) example to work with? Thanks.

**Comment by Andrzej-Andrzej** [comment]
My apologies for redirecting you, but because I have looked for help elsewhere too, here you are:
My desired output:
[https://community.rstudio.com/t/how-to-create-a-space-break-between-output-of-code-and-next-line-when-rendering-to-docx-in-quarto/165808](url)

and the code to create a table (flextable actually):
[https://stackoverflow.com/questions/76141089/how-to-merge-cells-in-flextable-multirow-header-in-flextable-in-r/76159499](url)

Any help is greatly appreciated, thank you.

**Reply by mcanouil** [reply]
Could you put and format the content here?
See <https://quarto.org/bug-reports.html>.

PS: be aware that cross-posting questions/problems is not considered a "good practice" or very "polite" by a lot of users, especially on Stack.

**Comment by Andrzej-Andrzej** [comment]
I am aware but still can't find a solution for this render behaviour anywhere or could a solutions to this be, that it is a bug ? 
How long should I wait for a solution until I look for help elsewhere (days, weeks) ?

**Reply by mcanouil** [reply]
I personally don't mind much about cross-posting, but choosing the right place to ask the question the first time matters.

This being said, none of your links are valid, and I personally don't have or want to spend time to look through several threads to figure out a possible (not) working example to work with in order to actually help.
So if you really want people to help, please help them first by providing all that is needed for them to actually help, this means for example setting up a repository with all relavant information, or put all documents/code in the post (not as attachment).

**Comment by Andrzej-Andrzej** [comment]
Yes you are right, I did not know about links not working, I do not know why (my apologies regarding that matter). I included links here only because attaching screenshots is not possible here. 
I have prepared reprex below, please find attached code of qmd file as requested:
````
---
title: "My_quarto"
format: docx

---


```{r}
library(flextable)
library(tidyverse)
library(officer)
library(magrittr)


M <- as.table(rbind(c(762, 327, 468), c(484, 239, 477)))

dimnames(M) <- list(
  gender = c("F", "M"),
  party = c("Democrat", "Independent", "Republican"))

M

f <- ftable(M,
            row.vars = "gender",
            col.vars = c("party"))


ftable_to_flextable <- function( x ){

  row.vars = attr( x, "row.vars" )
  col.vars = attr( x, "col.vars" )
  rows <- rev( expand.grid( rev(row.vars), stringsAsFactors = FALSE ) )
  cols <- rev(expand.grid( rev(col.vars), stringsAsFactors = FALSE ))

  xmat <- as.matrix(x)
  cols$col_keys = dimnames(xmat)[[2]]
  xdata <- cbind(
    data.frame(rows, stringsAsFactors = FALSE),
    data.frame(xmat, stringsAsFactors = FALSE)
  )
  names(xdata) <- c(names(row.vars), cols$col_keys)

  ft <- flextable(xdata)
  ft <- set_header_df(ft, cols)
  ft <- theme_booktabs(ft)
  ft <- merge_v(ft, j = names(row.vars))
  ft
}



final <- ftable(M) %>% ftable_to_flextable()

ft_2 <- add_header_row(final, values = c("Party"), colwidths = c(4), top = TRUE)

ft_2 <- add_header_row(final, values = c("Gender", "Party"), colwidths = c(1, 3), top = TRUE)

ft_2 <- width(ft_2, j=1, 3/2.54)

ft_2 <- align(ft_2, align = "center", part = "header")

ft_2 <- border_outer(ft_2, fp_border(color="black", width=1.5))

ft_2 <- border_inner(ft_2, fp_border(color="gray"))

ft_2 <- fix_border_issues(ft_2)

ft_2 %<>% bold(part = "header")

ft_2 %<>% bold (i = c(1,2),
               j = 1, bold = TRUE)

ft_2


```

```{r}

ft_2 %>% vline(i = c(1,2), j= 1,  fp_border(color="red", width=1.5))

ft_2 %>% border(border.bottom = fp_border(color="red", width=1.5), part = "header") 

ft_2 %<>% border(border.bottom = fp_border(color="red", width=1.5), part = "header") 

ft_2

```

````
[https://community.rstudio.com/t/how-to-create-a-space-break-between-output-of-code-and-next-line-when-rendering-to-docx-in-quarto/165808](https://community.rstudio.com/t/how-to-create-a-space-break-between-output-of-code-and-next-line-when-rendering-to-docx-in-quarto/165808)

**Reply by mcanouil** [reply]
A reprex is a small thing allowing us to reproduce.
For example

````qmd
---
title: "My_quarto"
format: docx
---

```{r}
library(flextable)
library(officer)
ft <- flextable(airquality[ sample.int(5),]) |>
  add_header_row(colwidths = c(4, 2), values = c("Air quality", "Time")) |>
  theme_vanilla() |>
  border_outer(fp_border(color = "black", width = 1.5)) |>
  border_inner(fp_border(color = "gray"))
ft
```

```{r}
ft
```

````

It produces 
<img width="524" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/a02786f0-2f4b-40b6-9780-8fc19fbc8050">

There are three ways to handle styling in Word document:
1. Manually
2. By defining a reference docx, see <https://quarto.org/docs/output-formats/ms-word-templates.html>
3. By defining some fenced/spans divs or using existing one in addition to a custom Lua filter

**Reply by mcanouil** [reply]
I am afraid, there is no easy way when it comes to Word.
FYI, the option in Word is "distance from surrounding text" in "table properties".

Alternatively, you can hack your way around this by manually adding `&nbsp;` between the two chunks or by creating a `knitr` hook to add it (<https://bookdown.org/yihui/rmarkdown-cookbook/output-hooks.html>).

````qmd
---
title: "Demo"
format: docx
---

```{r}
library(flextable)
library(officer)
ft <- flextable(airquality[ sample.int(5),]) |>
  add_header_row(colwidths = c(4, 2), values = c("Air quality", "Time")) |>
  theme_vanilla() |>
  border_outer(fp_border(color = "black", width = 1.5)) |>
  border_inner(fp_border(color = "gray"))
ft
```

&nbsp;

```{r}
ft
```
````

<img width="527" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/2b0ef072-1a75-449f-8b44-7892328902bd">

**Reply by Andrzej-Andrzej** [reply]
If I separate it into independent chunks there is no problem, I want to do it when everything is placed in one chunk, I mean:
code,
table,
code,
table, and so on. I do not want manually edit in Word if possible, as because in a long report where many tables are created it would be tedious. By the way how do you attach screenshots here ?

**Reply by mcanouil** [reply]
Either by using the buttons or by pasting them.

Well, in my opinion, Word is the worst format for automatic reporting and styling, _i.e._, there is no style component to control the "distance from surrounding text" globally for all tables thus how can you expect other tools to fill the gap in Word?

As I said, you have to hack your way, or leave docx format.

A hack could to be to add an empty row without border or any style at the end of each tables.

**Comment by Andrzej-Andrzej** [comment]
Thank you. Is there such a thing like controlling "distance from surrounding text" globally for tables in other formats besides docs ?
Do you know by any chance how to remove gray background produced by Quarto rendering in Word documents surrounding a code, but leaving code formatting intact ? I can remove it all by "clear text format" option but this removes everything. I want only to remove gray background/rectangle surrounding that code in Word.

**Reply by mcanouil** [reply]
> Thank you. Is there such a thing like controlling "distance from surrounding text" globally for tables in other formats besides docs ?

As I said, I looked (maybe missed it) but did not find one.

> Do you know by any chance how to remove gray background produced by Quarto rendering in Word documents surrounding a code, but leaving code formatting intact ? I can remove it all by "clear text format" option but this removes everything. I want only to remove gray background/rectangle surrounding that code in Word.

This is controlled by the reference docx file, take it, edit it, use it, see 
<https://quarto.org/docs/output-formats/ms-word-templates.html>.

**Reply by mcanouil** [reply]
Also if you really need Word that much, you should consider using dedicated tools, such as `officedown` in R.

**Comment by Andrzej-Andrzej** [comment]
Thank you very much for your help and explanations. Much appreciated.
